homeassistant/components/alarm_control_panel/egardia.py

```diff
@@ -71,40 +75,57 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     rs_enabled = config.get(CONF_REPORT_SERVER_ENABLED)
     rs_port = config.get(CONF_REPORT_SERVER_PORT)
     rs_codes = config.get(CONF_REPORT_SERVER_CODES)
+    version = config.get(CONF_VERSION)
 
     try:
         egardiasystem = egardiadevice.EgardiaDevice(
-            host, port, username, password, '')
+            host, port, username, password, '', version)
     except requests.exceptions.RequestException:
         raise exc.PlatformNotReady()
     except egardiadevice.UnauthorizedError:
         _LOGGER.error("Unable to authorize. Wrong password or username")
         return False
 
+    if rs_enabled:
+        # Set up the egardia server
+        _LOGGER.info("Setting up EgardiaServer")
+        try:
+            if D_EGARDIASRV not in hass.data:
+                hass.data[D_EGARDIASRV] = egardiaserver.EgardiaServer('',
+                                                                      rs_port)
+                bound = hass.data[D_EGARDIASRV].bind()
+                if not bound:
+                    raise IOError("Binding error occurred while " +
+                                  "starting EgardiaServer")
+        except IOError:
+            return False
+
     add_devices([EgardiaAlarm(
-        name, egardiasystem, hass, rs_enabled, rs_port, rs_codes)], True)
+        name, egardiasystem, hass, rs_codes)], True)
 
 
 class EgardiaAlarm(alarm.AlarmControlPanel):
     """Representation of a Egardia alarm."""
 
-    def __init__(self, name, egardiasystem, hass, rs_enabled=False,
-                 rs_port=None, rs_codes=None):
+    def __init__(self, name, egardiasystem, hass,
+                 rs_codes=None):
         """Initialize object."""
         self._name = name
         self._egardiasystem = egardiasystem
-        self._status = STATE_UNKNOWN
-        self._rs_enabled = rs_enabled
-        self._rs_port = rs_port
+        self._status = None
+        self._egardiaserver = hass.data[D_EGARDIASRV]
         self._hass = hass
 
         if rs_codes is not None:
             self._rs_codes = rs_codes[0]
         else:
             self._rs_codes = rs_codes
 
-        if self._rs_enabled:
-            self.listen_to_system_status()
+        if self._egardiaserver is not None:
+            _LOGGER.debug("Starting EgardiaServer and registering callback")
+            # Register callback for alarm status changes through EgardiaServer
+            self._egardiaserver.register_callback(self.handle_status_event)
```

Comment:
You don't need to do this inside the entity. You have access to the callback method of the entity as soon as it has been instantiated. Please move the registration and start of the server to outside the entity in setup platform. Then you don't need to pass in `hass` to the entity either.